mqtt/mqtt-broker/src/edgehub_auth.rs

```diff
@@ -0,0 +1,370 @@
+        }
+
+        Ok(response)
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct EdgeHubAuthRequest {
+    version: String,
+    username: Option<String>,
+    password: Option<String>,
+    certificate: Option<String>,
+    certificate_chain: Option<Vec<String>>,
+}
+
+impl From<(Option<String>, Credentials)> for EdgeHubAuthRequest {
+    fn from(item: (Option<String>, Credentials)) -> Self {
+        let (password, certificate) = match item.1 {
+            Credentials::Password(p) => (p, None),
+            Credentials::ClientCertificate(c) => (None, Some(encode(c))),
+        };
+
+        EdgeHubAuthRequest {
+            version: API_VERSION.to_string(),
+            username: item.0,
+            password,
+            certificate,
+            certificate_chain: None,
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+#[repr(i32)]
+pub enum EdgeHubResultCode {
+    Authenticated = 200,
+    Unauthenticated = 403,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct EdgeHubAuthResponse {
+    version: String,
+    result: EdgeHubResultCode,
+    identity: Option<String>,
+}
+
+impl EdgeHubAuthResponse {
+    fn into_auth_id(self) -> Result<Option<AuthId>, Error> {
+        match self.result {
+            EdgeHubResultCode::Authenticated => self
+                .identity
+                .map_or(Err(ErrorKind::AuthProcessResponse.into()), |i| {
+                    Ok(Some(AuthId::Value(i)))
+                }),
+            EdgeHubResultCode::Unauthenticated => Ok(None),
+        }
+    }
+}
+
+macro_rules! impl_for{
+    ($fun_name:ident, $typ:ident) => {
+        fn $fun_name<E: serde::de::Error>(self, v: $typ) -> Result<EdgeHubResultCode, E> {
+            match v {
+                200 => Ok(EdgeHubResultCode::Authenticated),
+                403 => Ok(EdgeHubResultCode::Unauthenticated),
+                v => Err(E::custom(format!("unexpected int: {}", v))),
+            }
+        }
+    }
+}
+
+struct EdgeHubResultCodeVisitor;
+impl<'de> serde::de::Visitor<'de> for EdgeHubResultCodeVisitor {
+    type Value = EdgeHubResultCode;
+    fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("200 or 403")
+    }
+
+    impl_for!(visit_i32, i32);
+    impl_for!(visit_i64, i64);
+    impl_for!(visit_u32, u32);
+    impl_for!(visit_u64, u64);
+}
+
+impl<'de> serde::Deserialize<'de> for EdgeHubResultCode {
+    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<EdgeHubResultCode, D::Error> {
+        d.deserialize_i32(EdgeHubResultCodeVisitor)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use base64::decode;
+    use matches::assert_matches;
+    use mockito::{mock, Matcher};
+
+    use crate::auth::{AuthId, Authenticator, Certificate, Credentials};
+    use crate::edgehub_auth::EdgeHubAuthenticator;
+    use crate::Error;
+
+    const CERT: &str = "MIIBLjCB1AIJAOTg4Zxl8B7jMAoGCCqGSM49BAMCMB8xHTAbBgNVBAMMFFRodW1i\
```

Comment:
I hope it's not really cert and url below don't exist. For unit tests you can use fake values, right?